MA-2318: get DiscussionTopic against topic id

wajeeha-khalid opened the pull request:
## [MA-2318](https://openedx.atlassian.net/browse/MA-2318)

Updated GET course topic endpoint to take topic_id as optional parameter and return details of discussion topic

### Testing
- [x] Unit

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @jcdyer 
- [x] Code review: @verdanmahmood 

### Post-review
- [x] Squash commits

wajeeha-khalid commented:
jenkins run js

nasthagiri commented:
@cahrens @robrap Did one of you guys want to take ownership of reviewing discussion-related PRs, like this?

robrap commented:
I'm still a little fuzzy on the current need for this. Is there or is there not an existing endpoint that would work? What is new in mobile that the current APIs don't support? Is this only needed for deep links, and is that what we are adding? If my questions are not wholly relevant, then maybe these questions help clear up what I need here or in the ticket. Thank you.

wajeeha-khalid commented:
@robrap That is right. This requirement came up for deep links.
Previously there is one endpoint that retrieves all module topics against a course and those topics are grouped into categories and has children for inline discussion topics.
With this PR, we have updated the existing endpoint to handle addition optional parameter (i.e. topic_id) to retrieve discussion topic's information.

wajeeha-khalid commented:
@robrap @muzaffaryousaf Would you guys be able to git it a look today?

nasthagiri commented:
@wajeeha-khalid There is already a Discussion Topics endpoint that can be accessed with `/api/discussion/v1/course_topics/{course_id}/`.  Clients can use that endpoint to get more information on each topic.  If needed, you can add a query parameter to filter by `topic_id`.

wajeeha-khalid commented:
@nasthagiri Thats exactly what I have implemented in this PR. I have updated existing Discussion Topics endpoint with an additional optional parameter `topic_id`

robrap commented:
@wajeeha-khalid: Sorry about the confusion.  I'm not sure what I misread that made me think this was a new endpoint, and then I mislead @nasthagiri. I'll connect with you over chat.  Thanks.

wajeeha-khalid commented:
@robrap no problem :)

robrap commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -244,6 +248,20 @@ def get_sorted_modules(category):
 
+def get_course_topic(request, course_key, topic_id):
+    """
+    Returns details of discussion topic against topic_id
```
against => for the given.

Also, should have Arguments, Returns, and Raises sections as usual.

robrap replied:
Probably needs some sort of TopicNotFoundError and a related test.

robrap commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -40,7 +40,11 @@
+from django_comment_client.utils import (
+    get_accessible_discussion_modules,
+    is_commentable_cohorted,
+    get_discussion_module,
```
Nit: alphabetical order.

robrap commented on lms/djangoapps/django_comment_client/utils.py:
```diff
@@ -137,6 +138,29 @@ def get_accessible_discussion_modules(course, user, include_all=False):  # pylin
     ]
 
 
+def get_discussion_module(course, user, discussion_id):
```
1. I was trying to understand what type of code reuse and/or additions might be needed by reviewing `discussion_category_id_access`, but this just got me more confused.  Do you know what a Category is?  Is it the same or different from a Topic?
2. I think we need to enhance our comments or something to better understand how and when a Topic ID can become a Discussion ID.  Are these really always the same thing?  Is there any difference?  For example, if these comments said "... a discussion id (a.k.a topic id)", would that be 100% accurate always?

robrap replied:
Some other points about `discussion_category_id_access` is that it checks if the cached key isn't set, it has a fallback, it has slightly different has_access code.  I just need help getting cleared up and I'm wondering if we could update docs to help clarify as well.  Thanks.

wajeeha-khalid replied:
@robrap For discussion modules of a course; category is the module type `e.g. 'Getting Started', 'Google Hangout', 'Getting Graded', 'Essays' etc.`

Each module category type could have multiple discussion targets e.g. 'Getting Started' have 'Working with Videos' where the discussion exists. At that level it is courseware discussion and the ID of that discussion is used to identify it and also referred as `topic_id` for that discussion. The name 'Working with Videos' also appears in topic dropdown selection in `forums` tab.

If 'General' topic type is selected to create a post from forums tab; we consider it as non-courseware topic.

Now; `platform` and `cs_comments_service` consider both `topic_id` and `discussion_id` as `commentable_id`.
For `a discussion id (a.k.a topic id); would that be 100% accurate always` I believe @nasthagiri could give a better insight on this.

wajeeha-khalid replied:
@robrap I was unaware of existence of `discussion_category_id_access`; and using this method instead of has_access would be more efficient as it uses cache.
I would need to refactor this method a little to make use of it. Will update.

jcdyer commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -40,7 +40,11 @@
+    get_discussion_module,
+    is_commentable_cohorted,
+)
 from lms.djangoapps.discussion_api.pagination import DiscussionAPIPagination
```
Can you move the other `discussion_api` imports (LL19-27) down here as well, and rename them as `lms.djangoapps.discussion_api`?

jcdyer commented on lms/djangoapps/django_comment_client/utils.py:
```diff
@@ -137,6 +138,33 @@ def get_accessible_discussion_modules(course, user, include_all=False):  # pylin
+            raise DiscussionNotFoundError("Discussion not found.")
+        return discussion_module
+    else:
+        raise DiscussionNotFoundError("Discussion not found.")
```
The message included with the exception doesn't provide any useful information.  It just duplicates the name of the exception.  Either leave it out, or include the id of the discussion that was not found.

jcdyer commented on openedx/core/lib/exceptions.py:
```diff
@@ -10,3 +10,8 @@ class CourseNotFoundError(ObjectDoesNotExist):
+
+class DiscussionNotFoundError(ObjectDoesNotExist):
+    """ Discussion Module was not found. """
+    pass
```
Nit: I see you're following the style of the rest of the docstrings in this file, but the extra spaces hurt my brain.  Can you switch to the three-line docstring style?

jcdyer commented on lms/djangoapps/discussion_api/views.py:
```diff
@@ -75,27 +76,41 @@ class CourseTopicsView(DeveloperErrorViewMixin, APIView):
-
-        * non_courseware_topics: The list of topic trees that are not linked to
-              courseware. Items are of the same format as in courseware_topics.
     """
```
I'm not sure I like having one point endpoint return responses with such different formats depending on an optional parameter.  It feels like a normal list/detail pattern (list of topics vs. individual topic), where the list should show up at `/path/to/list/` and the the detail at `/path/to/list/{detail_id}/`.  Having the same endpoint handle both cases makes it less clear for client code to figure out what format to expect.  With the parameter, the top-level of the response is a topic object, and without it, it's a dictionary of two lists. 

I know @robrap and @nasthagiri were pushing to reuse the same endpoint, and I don't want to just be contradictory, so I'd love to hear further discussion on this point.

robrap replied:
@jcdyer: I agree with your feedback.  To get more explicit about the new endpoint, do you think we should include `topic` to make it more clear?:
```
/api/discussion/v1/course_topics/course-v1:ExampleX+Subject101+2015/topic/a29d117faf43c7c59eda7717e722e9d8f4a47ea8
```
Or follow your format of just adding the topic id?:
```
/api/discussion/v1/course_topics/course-v1:ExampleX+Subject101+2015/a29d117faf43c7c59eda7717e722e9d8f4a47ea8
```
I like the former, including `topic`, but wondering what others think.

jcdyer replied:
@robrap I could be persuaded either way.  topic is already mentioned in the `/course_topics/`segment of the URL, and I tend to like URLs that make sense if you take away segments from the end, and 

    /api/discussion/v1/course_topics/course-v1:ExampleX+Subject101+2015/topic/

doesn't represent anything real.  On the other hand, 

    /api/discussion/v1/course_topics/course-v1:ExampleX+Subject101+2015/a29d117faf43c7c59eda7717e722e9d8f4a47ea8

is pretty hard to decipher.

I don't think edX has an established goal of making all URL segments meaningful, though, so unless someone else has a strong opinion, let's go with your first example.

jcdyer replied:
@wajeeha-khalid There was further discussion about the above in the API room on HipChat, but to summarize, I think what we'd like to see is that we use the same URL structure you've already implemented, with topic_id as a keyword parameter, but keep the same response structure as when the param is missing, so that when someone requests:

    /api/discussion/v1/course_topics/course-v1:ExampleX+Subject101+2015/?topic_id=a29d117faf43c7c59eda7717e722e9d8f4a47ea8

they will get a response that looks like:

    {"courseware_topics": [{<topic>}], "non_courseware_topics": []}

or

    {"courseware_topics": [], "non_courseware_topics": [{<topic>}]}

Additionally, can you make topic_id a MultiValueField, so the user can request more than one topic_id in a single request (to make it consistent with other discussion endpoints)?

wajeeha-khalid commented:
jenkins run python
jenkins run bokchoy

wajeeha-khalid commented:
@jcdyer @verdanmahmood Updated PR with feedback and also made topic_id to accept list of ids. Kindly give it a second pass.

jcdyer commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -177,103 +185,136 @@ def get_course(request, course_key):
+                    module.discussion_target,
+                    get_thread_list_url(request, course_key, [module.discussion_id]),
+                )
+                children.append(discussion_topic.__dict__)
```
Do we have a better way to serialize discussion topics?  Like maybe a DRF Serializer subclass?  This seems like it could introduce unexpected data (internal instance variables, for example).

jcdyer commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -177,103 +185,136 @@ def get_course(request, course_key):
+    )
+
+    if topic_ids:
+        not_found_topic_ids = list(set(topic_ids) - (
```
Shouldn't need to convert this back to a list.  Let not_found_topic_ids remain a set.

jcdyer commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -177,103 +185,136 @@ def get_course(request, course_key):
+            set(existing_courseware_topic_ids) | set(existing_non_courseware_topic_ids)
+        ))
+        if not_found_topic_ids:
+            raise DiscussionNotFoundError("Discussion not found for '{}'.".format(",".join(not_found_topic_ids)))
```
for readability of error messages, join with `", "` instead of `","`.

jcdyer commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -177,103 +185,136 @@ def get_course(request, course_key):
     """
-    Returns details of discussion topic for the given topic_id.
+    non_courseware_topics = []
+    existing_topic_ids = []
```
topic_ids is used for testing membership on every pass through the loop.  Make it a set for better performance.

jcdyer replied:
`existing_topic_ids` is also used as a set when it gets returned.  Make it a set here, and save the conversion later.

jcdyer commented on lms/djangoapps/discussion_api/views.py:
```diff
@@ -76,41 +75,30 @@ class CourseTopicsView(DeveloperErrorViewMixin, APIView):
     **Example Requests**:
```
Rename this section: `**Request Format**`, since it a verbatim example.

wajeeha-khalid replied:
@jcdyer Is not **Example Requests** being followed through edx? The `discussion_api` is following it and I see same convention in `course_block_api` and `mobile_api` too.

jcdyer replied:
Yeah.  I guess you're right.  Leave it alone, then.

jcdyer commented on openedx/core/lib/exceptions.py:
```diff
@@ -1,17 +1,25 @@
-    """ Discussion Module was not found. """
+    """
+    Discussion Module was not found.
+    """
```
Thanks!

jcdyer commented:
Looking better.  My one suggestion is to think about using sets more fluently as a basic data structure.  Not everything needs to be a list.  

In particular, sets are better when you need to test for membership (`x in set` happens in O(1) time, `x in list` takes O(n)), so long as ordering and repetition are not important.

wajeeha-khalid commented:
jenkins run bokchoy

verdanmahmood commented:
👍

jcdyer commented:
:+1:
